application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/DefaultLicensedExtensionManager.java

```diff
@@ -116,23 +121,69 @@
     }
 
     @Override
-    public Collection<ExtensionId> getMandatoryLicensedExtensions()
+    public Set<ExtensionId> getMandatoryLicensedExtensions()
+    {
+        Set<ExtensionId> mandatoryLicensedExtensions = computeMandatoryLicensedExtensions();
+        this.cachedMandatoryLicensedExtensions = mandatoryLicensedExtensions;
+        return mandatoryLicensedExtensions;
```

Comment:
I think we can avoid synchronization if the mandatory licensed extensions are already cached.

```suggestion
        Set<ExtensionId> mandatoryLicensedExtensions = this.cachedMandatoryLicensedExtensions;
        if (mandatoryLicensedExtensions == null) {
            mandatoryLicensedExtensions = computeMandatoryLicensedExtensions();
            this.cachedMandatoryLicensedExtensions = mandatoryLicensedExtensions;
        }
        return Collections.unmodifiableSet(mandatoryLicensedExtensions);
```